packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs

```diff
@@ -0,0 +1,258 @@
+    ///
+    /// This function may return an `Error` variant if there is a problem with processing the block
+    /// proposal, updating the core info, processing raw state transitions, or processing block fees.
+    ///
+    pub(super) fn run_block_proposal_v0(
+        &self,
+        block_proposal: block_proposal::v0::BlockProposal,
+        transaction: &Transaction,
+    ) -> Result<ValidationResult<block_execution_outcome::v0::BlockExecutionOutcome, Error>, Error>
+    {
+        // Start by getting information from the state
+        let state = self.state.read().unwrap();
+
+        let last_block_time_ms = state.last_block_time_ms();
+        let last_block_height =
+            state.known_height_or(self.config.abci.genesis_height.saturating_sub(1));
+        let last_block_core_height =
+            state.known_core_height_or(self.config.abci.genesis_core_height);
+        let hpmn_list_len = state.hpmn_list_len();
+        let _quorum_hash = state.current_validator_set_quorum_hash;
+
+        let mut block_platform_state = state.clone();
+
+        // Init block execution context
+        let block_state_info = block_state_info::v0::BlockStateInfo::from_block_proposal(
+            &block_proposal,
+            last_block_time_ms,
+        );
+
+        // First let's check that this is the follower to a previous block
+        if !block_state_info.next_block_to(last_block_height, last_block_core_height)? {
+            // we are on the wrong height or round
+            return Ok(ValidationResult::new_with_error(AbciError::WrongBlockReceived(format!(
+                "received a block proposal for height: {} core height: {}, current height: {} core height: {}",
+                block_state_info.height, block_state_info.core_chain_locked_height, last_block_height, last_block_core_height
+            )).into()));
+        }
+
+        // destructure the block proposal
+        let block_proposal::v0::BlockProposal {
+            consensus_versions: _,
+            block_hash: _,
+            height,
+            round: _,
+            core_chain_locked_height,
+            proposed_app_version,
+            proposer_pro_tx_hash,
+            validator_set_quorum_hash,
+            block_time_ms,
+            raw_state_transitions,
+        } = block_proposal;
+        // todo: verify that we support the consensus versions
+        // We start by getting the epoch we are in
+        let genesis_time_ms = self.get_genesis_time_v0(height, block_time_ms, transaction)?;
+
+        let epoch_info =
+            EpochInfo::from_genesis_time_and_block_info(genesis_time_ms, &block_state_info)?;
+
+        let block_info = block_state_info.to_block_info(
+            Epoch::new(epoch_info.current_epoch_index)
+                .expect("current epoch index should be in range"),
+        );
+
+        // Update the masternode list and create masternode identities and also update the active quorums
+        self.update_core_info_v0(
```

Comment:
It is not a hack, there might be slightly different versions of the same function, and we need to be able make sure that a specific version stays present. While it might look not as good as some other solutions I tried out, I believe this is the safest.